Userland/Libraries/LibArchive/Lzma.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023, Tim Schumacher <timschumi@gmx.de>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibArchive/Lzma.h>
+
+namespace Archive {
+
+u32 LzmaHeader::dictionary_size() const
+{
+    // "If the value of dictionary size in properties is smaller than (1 << 12),
+    //  the LZMA decoder must set the dictionary size variable to (1 << 12)."
+    constexpr u32 minimum_dictionary_size = (1 << 12);
+    if (m_dictionary_size < minimum_dictionary_size)
+        return minimum_dictionary_size;
+
+    return m_dictionary_size;
+}
+
+Optional<u64> LzmaHeader::uncompressed_size() const
+{
+    // We are making a copy of the packed field here because we would otherwise
+    // pass an unaligned reference to the constructor of Optional, which is
+    // undefined behavior.
+    auto uncompressed_size = m_uncompressed_size;
+
+    // "If "Uncompressed size" field contains ones in all 64 bits, it means that
+    //  uncompressed size is unknown and there is the "end marker" in stream,
+    //  that indicates the end of decoding point."
+    if (uncompressed_size == UINT64_MAX)
+        return {};
+
+    // "In opposite case, if the value from "Uncompressed size" field is not
+    //  equal to ((2^64) - 1), the LZMA stream decoding must be finished after
+    //  specified number of bytes (Uncompressed size) is decoded. And if there
+    //  is the "end marker", the LZMA decoder must read that marker also."
+    return uncompressed_size;
+}
+
+ErrorOr<void> LzmaHeader::decode_model_properties(u8& literal_context_bits, u8& literal_pos_bits, u8& pos_bits) const
+{
+    // "Decodes the following values from the encoded model properties field:
+    //
+    //     name  Range          Description
+    //       lc  [0, 8]         the number of "literal context" bits
+    //       lp  [0, 4]         the number of "literal pos" bits
+    //       pb  [0, 4]         the number of "pos" bits
+    //
+    //  Encoded using `((pb * 5 + lp) * 9 + lc)`."
+
+    u8 input_bits = m_model_properties;
+
+    if (input_bits >= (9 * 5 * 5))
+        return Error::from_string_literal("Encoded model properties value is larger than the highest possible value");
+
+    literal_context_bits = input_bits % 9;
+    input_bits /= 9;
+    VERIFY(literal_context_bits >= 0 && literal_context_bits <= 8);
```

Comment:
Just to confirm: These properties on the bits are mandated by either the math here in this function or by the caller of this function? (i.e. we won't see any fuzzer crashes hitting these asserts, right?)